crypto/rand.go

```diff
@@ -2,109 +2,103 @@ package crypto
 
 import (
 	"crypto/rand"
+	"hash"
+	"io"
+	"math"
 	"math/big"
-	"sync/atomic"
+	"sync"
+	"unsafe"
 )
 
-var (
-	// Two counters is used in case the first overflows. The first is not at
-	// risk of overflowing, but we're paranoid.
-	counter     uint64
-	counter2    uint64
-	entropyBase Hash
-)
-
-// init will create an entropy pool that the RNG will draw from. On most
-// systems, hashing out of a preset RNG pool will be substantially faster than
-// using crypto/rand, which relies on syscalls.
-func init() {
-	// Use 64 bytes in case the first 32 aren't completely random.
-	base := make([]byte, 64)
-	_, err := rand.Read(base)
-	if err != nil {
-		panic("unable to set entropy for the crypto package")
-	}
-	entropyBase = HashObject(base)
+// A randReader produces random values via repeated hashing. The entropy field
+// is the concatenation of an initial seed and a 128-bit counter. Each time
+// the entropy is hashed, the counter is incremented.
+type randReader struct {
+	entropy [16 + HashSize]byte
+	h       hash.Hash
+	buf     [32]byte
+	mu      sync.Mutex
 }
```

Comment:
There is no reason for this object. It's not something anyone should ever have to customize. Random numbers are random numbers, what goes on under the hood doesn't matter as long as what comes out is secure.